docs/syntax-if-statements-initializers.md

```diff
@@ -0,0 +1,94 @@
+# Initializers in if statements
+
+# Summary
+
+Introduce an initializer expression that declares and initializes variables in if statements.
+
+# Motivation
+
+Initializers can improve code clarity and reduce scope pollution by allowing developers to declare variables in the conditions of if statements. ⁤⁤ Declaring variables at the point of use in if statements for the scope of the if statement's block simplifies code, leading to better readability and understanding of program logic. By limiting the scope of variables to the condition's pertaining block, the risk of unintended variable reuse and naming conflicts is reduced. ⁤
+
+The reduced scope pollution improves register space in extreme cases (or auto-generated code) where developers have many variables defined and have to work around register limits by reducing the register size. In some scenarios, especially on Roblox, initializing variables within if statements can lead to improved performance and reduced complexity by avoiding unnecessary calls by developers. ⁤ A common paradigm used by Roblox developers is to use `Instance:FindFirstChild` in their condition and then access it afterwards instead of keeping an existing variable around in the new scope, polluting the existing scope.
+
+# Design
+
+If statements with initializers must match (following the Luau grammar) `'if' 'local' bindinglist ['=' explist] 'then'` and `'local' bindinglist ['=' explist] where exp 'then'` syntax. Parenthesis are also allowed around the initializer for consistency with other expressions. The variables declared by the initializer are only available to the block of that condition and will be undefined to the `elseif` conditions and blocks as well as the `else` block; any code after the if statement won't have the variables defined either.
```

Comment:
> Parenthesis are also allowed around the initializer for consistency with other expressions.

Can you give an example?

Reply:
> Can you give an example? I'm not sure what you mean here and I hope it's not `if (local a = foo()) then`

That is exactly what it means, I felt it would be awkward to not allow `if (local a = foo()) where (a == 15) then`.

Reply:
That will be a 'no' from me.

Reply:
The syntax would be inconsistent but it can be changed if it is somehow a problem, I feel it might be weird if you're writing out a condition and the initializer itself cant be parenthesized.

Reply:
It will actually be consistent, `()` is part of the expression grammar, local is a part of the if statement grammar.
The ability to write `if (a == 2)` like in other languages is just a coincidence of the ability to write group expressions.

Reply:
> It will actually be consistent if it's removed; `()` is part of the expression grammar, local is a part of the if statement grammar. 

I guess that would also be fitting, I updated the RFC to remove that.